xarray/tests/test_concat.py

```diff
@@ -50,8 +135,518 @@ def test_concat_compat() -> None:
+    #   the resulting set
+
+    ds_result = Dataset(
+        data_vars={
+            # pressure will be first in this since the first dataset is missing this var
+            # and there isn't a good way to determine that this should be first
+            # this also means temperature will be last as the first data vars will
+            # determine the order for all that exist in that dataset
+            "pressure": (["x", "y", "day"], result_vars["pressure"]),
+            "humidity": (["x", "y", "day"], result_vars["humidity"]),
+            "precipitation": (["x", "y", "day"], result_vars["precipitation"]),
+            "cloud cover": (["x", "y", "day"], result_vars["cloud cover"]),
+            "temperature": (["x", "y", "day"], result_vars["temperature"]),
+        },
+        coords={
+            "lat": (["x", "y"], datasets[0].lat.values),
+            "lon": (["x", "y"], datasets[0].lon.values),
+            #  "day": ["day" + str(d + 1) for d in range(2 * len(vars_to_drop))],
+        },
+    )
+
+    result = concat(datasets, dim="day")
+    r1 = list(result.data_vars.keys())
+    r2 = list(ds_result.data_vars.keys())
+    assert r1 == r2  # check the variables orders are the same
+
+    assert_equal(result, ds_result)
+
+
+def test_multiple_datasets_with_missing_variables() -> None:
+    vars_to_drop = [
+        "temperature",
+        "pressure",
+        "humidity",
+        "precipitation",
+        "cloud cover",
+    ]
+    datasets = create_concat_datasets(len(vars_to_drop), 123)
+    # set up the test data
+    datasets = [datasets[i].drop_vars(vars_to_drop[i]) for i in range(len(datasets))]
+
+    # set up the validation data
+    # the below code just drops one var per dataset depending on the location of the
+    # dataset in the list and allows us to quickly catch any boundaries cases across
+    # the three equivalence classes of beginning, middle and end of the concat list
+    result_vars = dict.fromkeys(vars_to_drop)
+    for i in range(len(vars_to_drop)):
+        for d in range(len(datasets)):
+            if d != i:
+                if result_vars[vars_to_drop[i]] is None:
+                    result_vars[vars_to_drop[i]] = datasets[d][vars_to_drop[i]].values
+                else:
+                    result_vars[vars_to_drop[i]] = np.concatenate(
+                        (
+                            result_vars[vars_to_drop[i]],
+                            datasets[d][vars_to_drop[i]].values,
+                        ),
+                        axis=2,
+                    )
+            else:
+                if result_vars[vars_to_drop[i]] is None:
+                    result_vars[vars_to_drop[i]] = np.full([1, 4, 2], np.nan)
+                else:
+                    result_vars[vars_to_drop[i]] = np.concatenate(
+                        (result_vars[vars_to_drop[i]], np.full([1, 4, 2], np.nan)),
+                        axis=2,
+                    )
```

Comment:
```
 xarray/tests/test_concat.py:412: error: Argument 1 to "concatenate" has incompatible type "Tuple[Optional[Any], ndarray[Any, dtype[Any]]]"; expected "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]]]"  [arg-type]
```
I think this means you can't concatenate None, `np.concatenate((None, [2]))`.
`result_vars = dict.fromkeys(vars_to_drop) ` starts with None values, and then later becomes np.ndarray. This means `result_vars: dict[str, None | np.ndarray]`, so you have to narrow the typing and remove that None somehow before np.concatenate. 

I was thinking a clear way is to initialize with a proper dict[str, np.ndarray]:
```python
    vars_to_drop: dict[str, np.ndarray] = {
        "temperature": np.array(),
        "pressure": np.array(),
        "humidity": np.array(),
        "precipitation": np.array(),
        "cloud cover": np.array(),
    }
```

`vars_to_drop` was a list before so you might have to change the code a little with `vars_to_drop.keys()`.

There may be a more performant solution as well but I can't think of it right now. 

Another idea; is that fromkeys important? Instead of:
```python
result_vars = dict.fromkeys(vars_to_drop)
```
you can try 
```python
result_vars: dict[str, np.ndarray] = {}
```

I think the rest are variants of this.

Reply:
Thanks a bunch, that makes a lot of sense, @Illviljan. I'm slowly getting behind this typing stuff.